session/tidb.go

```diff
@@ -139,62 +140,25 @@ func Compile(ctx context.Context, sctx sessionctx.Context, stmtNode ast.StmtNode
 	return stmt, errors.Trace(err)
 }
 
-// runStmt executes the sqlexec.Statement and commit or rollback the current transaction.
-func runStmt(ctx context.Context, sctx sessionctx.Context, s sqlexec.Statement) (sqlexec.RecordSet, error) {
-	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
-		span1 := span.Tracer().StartSpan("session.runStmt", opentracing.ChildOf(span.Context()))
-		span1.LogKV("sql", s.OriginText())
-		defer span1.Finish()
-	}
-
-	var err error
-	var rs sqlexec.RecordSet
-	se := sctx.(*session)
-	rs, err = s.Exec(ctx)
-	sessVars := se.GetSessionVars()
-	// All the history should be added here.
-	sessVars.TxnCtx.StatementCount++
-	if !s.IsReadOnly() {
-		if err == nil {
-			GetHistory(sctx).Add(0, s, se.sessionVars.StmtCtx)
-		}
-		if sctx.Txn(false).Valid() {
-			if err != nil {
-				sctx.StmtRollback()
-			} else {
-				sctx.StmtCommit()
-			}
-		}
-	}
-
-	// There are two known cases that the s.txn.fail is not nil:
-	// 1. active transaction fail, can't get start ts for example
-	// 2. transaction too large and StmtCommit fail
-	// On both cases, we can return error in advance.
-	if se.txn.fail != nil {
-		err = se.txn.fail
-		se.txn.cleanup()
-		se.txn.fail = nil
-		return nil, errors.Trace(err)
-	}
-
+func finishStmt(ctx context.Context, sctx sessionctx.Context, se *session, sessVars *variable.SessionVars, meetsErr error) error {
+	err := meetsErr
 	if !sessVars.InTxn() {
-		if err != nil {
+		if meetsErr != nil {
 			log.Info("RollbackTxn for ddl/autocommit error.")
 			err1 := se.RollbackTxn(ctx)
 			terror.Log(errors.Trace(err1))
 		} else {
 			err = se.CommitTxn(ctx)
 		}
-	} else {
+	} else if meetsErr == nil {
```

Comment:
what about:
```
if meetsErr != nil {
   return err
}

xxxx
```

Reply:
No. The logic is 
```
if !sessVars.InTxn() {
    ...
} else {
    if meetsErr == nil 
}
```

rather than
```
if !sessVars.InTxn() {
   ...
}
if meetsErr == nil {
}
```

Reply:
I've take @lysu 's advice